olive/passes/onnx/split.py

```diff
@@ -0,0 +1,280 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+from collections import defaultdict
+from copy import deepcopy
+from pathlib import Path
+from typing import Any, Dict
+
+import numpy as np
+import onnx
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import CompositeModelHandler, ONNXModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.onnx.common import get_external_data_config, model_proto_to_olive_model
+from olive.passes.onnx.onnx_dag import OnnxDAG
+from olive.passes.pass_config import PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class SplitModel(Pass):
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "include_all_nodes": PassConfigParam(
```

Comment:
It would make sense to make the decision of splitting transformer layers (i.e. layer block) only or ALL layers at the time of capturing split info.

Reply:
this is not straightforward. without a way to know how deep the split needs to be and how to handle nested layers, if we split all layers in the model, then it only gets split at a higher level. for example, the hierachy of layers in llama is 
- model
  - embed_tokens
  - layers
  - norm
- lm_head 

just splitting at the top level doesn't split the transformer layers since it's nested. this option is just to ensure that we get a list of models that can be chained together to be equivalent to the unsplit model.

Reply:
Could we hoist "include_all_nodes" into CaptureSplitInfo pass instead of here in SplitModel pass?

Reply:
yes, I can do that. make it an option in the capture split info part and save it as additional model metadata.

Reply:
In that case, leave it as it is.